README.md
# JFramework Command Interface

A Joomla Framework Command package that support nested commend calling.

## The Command Calling Flow

``` bash
$ php cli/app.php command1 command2 command3 -a -b -cd --help
```

``` text
command1
    ->configure
    ->execute
    
    commend2
        ->configure
        ->execute
        
        commend3
            ->configure
            ->execute
        
        return
    
    return

return
```

## Usage

We can using `Command` without `CliApplicaion`, just create this object.

### Simple One Command

``` php
// cli/app.php

use Joomla\Command\Command;
use Joomla\Command\Option;
use Joomla\Input\Cli as Input;

$command = new Command('app', new Input);

$command->setDescription('This is first level command description')
	->addOption(
		'q', // option name
		0,   // default value
		'Add this option can make output lower case.', // option description
		Option::IS_GLOBAL // sub command will extends this global option
	)
	->addOption(
		['y', 'yell', 'Y'], // First element will be option name, others will be alias
		0,
		'Yell will make output upper case.',
		Option::IS_NORMAL // sub command will not extends normal option
	)
	->setCode(
		function($command, $input, $output)
		{
			if (empty($input->args[0]))
			{
				$output->out('Please enter a name: ');
				$name = fread(STDIN, 8792);
			}
			else
			{
				$name = $input->args[0];
			}

			$reply = 'Hello ' . $name;

			if ($command->getOption('y'))
			{
				$reply = strtoupper($reply);
			}

			if ($command->getOption('q'))
			{
				$reply = strtolower($reply);
			}

			$output->out($reply);
		}
	);

$command->execute();
```

The `execute()` method will execute your command code. We set a `Closure` into `Command` and execute it.

When we type:

``` bash
$ php cli/app.php Asika
```

Output

``` text
Hello Asika
```

-----

Type:

``` bash
$ php cli/app.php Asika -q
```

Output

``` text
hello asika
```
-----

Type:

``` bash
$ php cli/app.php Asika -yell
```

Output

``` text
HELLO ASIKA
```

### Nested Command

``` php
// cli/app.php

use Joomla\Command\Command;
use Joomla\Command\Option;
use Joomla\Input\Cli as Input;

$command = new Command('app', new Input);

// Default Command
$command->setDescription('This is first level command description')
	->addOption(
		'q',
		0,
		'Add this option can make output lower case.',
		Option::IS_GLOBAL
	)
	// First level code
	->setCode(
		function($command, $input, $output)
		{
			$output->out('First level command.');
		}
	)
	// Second level commend
	->addArgument(
		'second',
		'The second level argument',
		array(
			new Option(
				array('a', 'A', 'ask'),
				'a default',
				'a desc'
			)
		),
		function($command, $input, $output)
		{
			echo 'Second level commend.';
		}
	);

$command->execute();
```

Type:

``` bash
$ php cli/app.php
```

Output

``` text
First level command.
```

-----

Type:

``` bash
$ php cli/app.php second
```

Output

``` text
Second level commend.
```

### Extends Class

If you want to execute Command by your class, you can extends it from `Command`:

``` php
<?php
// First level: src/Myapp/FooCommand.php

namespace Myapp;

use Joomla\Command\Command;

class FooCommand extends Command
{
	public function configure()
	{
		$this->setName('foo')
		    ->setDescription('foo desc')
			->addArgument(new Foo\BarCommand)
			->addOption(
			    'q',
			    0,
			    'q desc'
			)
			;
	}

	public function doExecute()
	{
		$this->output->out('Foo');
	}
}
```

And we add a sub command.

``` php
<?php
// Second level: src/Myapp/Foo/BarCommand.php

namespace Myapp\Foo;

use Joomla\Command\Command;

class BarCommand extends Command
{
	public function configure()
	{
		$this->setName('bar')
            ->setDescription('bar desc')
			;
	}

	public function doExecute()
    {
        $this->output->out('Bar');
    }
}
```

Type:

``` bash
$ php cli/app.php foo
```

Output

``` text
Foo
```

-----

Type:

``` bash
$ php cli/app.php foo bar
```

Output

``` text
Bar
```


docs/history.md
### Release History

#### Version 12.3

Version 12.3 ("Curiosity") of the Joomla Platform was tagged and released on 19 December 2012. 
[Release notes for version 12.3](http://developer.joomla.org/news/547-version-12-3-of-the-joomla-platform-released.html)

#### Version 12.2

Version 12.2 ("Neil Armstrong") of the Joomla Platform was tagged and released on 21 September 2012. 
[Release notes for version 12.2](http://developer.joomla.org/news/542-version-12-2-of-the-joomla-platform-released.html)

#### Version 12.1

Version 12.1 ("Louis Landry") of the Joomla Platform was tagged and
released on 10 May 2012.
[Release notes for version 12.1](http://developer.joomla.org/news/421-version-12-1-of-the-joomla-platform-released.html)

#### Version 11.4

Version 11.4 ("Brian Kernighan") of the Joomla Platform was tagged and
released on 4 January 2012.
[Release notes for version 11.4](http://developer.joomla.org/news/380-version-11-4-of-the-joomla-platform-released.html)

#### Version 11.3

Version 11.3 ("Dennis Ritchie") of the Joomla Platform was tagged and
released on 24 November 2011.
[Release notes for version 11.3](http://developer.joomla.org/news/378-version-11-3-of-the-joomla-platform-released.html)

#### Version 11.2

Version 11.2 ("Omar") of the Joomla Platform was tagged and released on
26 July 2011.
[Release notes for version 11.2](http://developer.joomla.org/news/363-version-11-2-of-the-joomla-platform-released.html)

#### Version 11.1

Version 11.1 was tagged on 5 July 2011. This was the inaugural release of
the platform following its separation from the Joomla CMS trunk.
[Release notes for version 11.1](http://developer.joomla.org/news/362-version-11-1-of-the-joomla-platform-released.html)


src/loader.php
<?php
/**
 * @package    Joomla\Framework
 *
 * @copyright  Copyright (C) 2005 - 2013 Open Source Matters, Inc. All rights reserved.
 * @license    GNU General Public License version 2 or later; see LICENSE
 */


/**
 * Static class to handle loading of libraries.
 *
 * @package  Joomla\Framework
 * @since    11.1
 */
abstract class JLoader
{	
	/**
	 * Container for already imported library paths.
	 *
	 * @var    array
	 * @since  11.1
	 */
	protected static $classes = array();

	/**
	 * Container for already imported library paths.
	 *
	 * @var    array
	 * @since  11.1
	 */
	protected static $imported = array();

	/**
	 * Container for registered library class prefixes and path lookups.
	 *
	 * @var    array
	 * @since  12.1
	 */
	protected static $prefixes = array();

	/**
	 * Container for namespace => path map.
	 *
	 * @var    array
	 * @since  12.3
	 */
	protected static $namespaces = array();

	/**
	 * Method to get the list of registered classes and their respective file paths for the autoloader.
	 *
	 * @return  array  The array of class => path values for the autoloader.
	 *
	 * @since   11.1
	 */
	public static function getClassList()
	{
		return self::$classes;
	}

	/**
	 * Method to get the list of registered namespaces.
	 *
	 * @return  array  The array of namespace => path values for the autoloader.
	 *
	 * @since   12.3
	 */
	public static function getNamespaces()
	{
		return self::$namespaces;
	}

	/**
	 * Loads a class from specified directories.
	 *
	 * @param   string  $key   The class name to look for (dot notation).
	 * @param   string  $base  Search this directory for the class.
	 *
	 * @return  boolean  True on success.
	 *
	 * @since   11.1
	 */
	public static function import($key, $base = null)
	{
		// Only import the library if not already attempted.
		if (!isset(self::$imported[$key]))
		{
			// Setup some variables.
			$success = false;
			$parts = explode('.', $key);
			$class = array_pop($parts);
			$base = (!empty($base)) ? $base : dirname(__DIR__) . '/vendor';
			$path = str_replace('.', DIRECTORY_SEPARATOR, $key);

			// Handle special case for helper classes.
			if ($class == 'helper')
			{
				$class = ucfirst(array_pop($parts)) . ucfirst($class);
			}
			// Standard class.
			else
			{
				$class = ucfirst($class);
			}

			// If we are importing a library from the Joomla namespace set the class to autoload.
			if (strpos($path, 'joomla') === 0)
			{
				// Since we are in the Joomla namespace prepend the classname with J.
				$class = 'J' . $class;

				// Only register the class for autoloading if the file exists.
				if (is_file($base . '/' . $path . '.php'))
				{
					self::$classes[strtolower($class)] = $base . '/' . $path . '.php';
					$success = true;
				}
			}
			/*
			 * If we are not importing a library from the Joomla namespace directly include the
			* file since we cannot assert the file/folder naming conventions.
			*/
			else
			{
				// If the file exists attempt to include it.
				if (is_file($base . '/' . $path . '.php'))
				{
					$success = (bool) include_once $base . '/' . $path . '.php';
				}
			}

			// Add the import key to the memory cache container.
			self::$imported[$key] = $success;
		}

		return self::$imported[$key];
	}

	/**
	 * Load the file for a class.
	 *
	 * @param   string  $class  The class to be loaded.
	 *
	 * @return  boolean  True on success
	 *
	 * @since   11.1
	 */
	public static function load($class)
	{
		// Sanitize class name.
		$class = strtolower($class);

		// If the class already exists do nothing.
		if (class_exists($class, false))
		{
			return true;
		}

		// If the class is registered include the file.
		if (isset(self::$classes[$class]))
		{
			include_once self::$classes[$class];

			return true;
		}

		return false;
	}

	/**
	 * Directly register a class to the autoload list.
	 *
	 * @param   string   $class  The class name to register.
	 * @param   string   $path   Full path to the file that holds the class to register.
	 * @param   boolean  $force  True to overwrite the autoload path value for the class if it already exists.
	 *
	 * @return  void
	 *
	 * @since   11.1
	 */
	public static function register($class, $path, $force = true)
	{
		// Sanitize class name.
		$class = strtolower($class);

		// Only attempt to register the class if the name and file exist.
		if (!empty($class) && is_file($path))
		{
			// Register the class with the autoloader if not already registered or the force flag is set.
			if (empty(self::$classes[$class]) || $force)
			{
				self::$classes[$class] = $path;
			}
		}
	}

	/**
	 * Register a class prefix with lookup path.  This will allow developers to register library
	 * packages with different class prefixes to the system autoloader.  More than one lookup path
	 * may be registered for the same class prefix, but if this method is called with the reset flag
	 * set to true then any registered lookups for the given prefix will be overwritten with the current
	 * lookup path. When loaded, prefix paths are searched in a "last in, first out" order.
	 *
	 * @param   string   $prefix  The class prefix to register.
	 * @param   string   $path    Absolute file path to the library root where classes with the given prefix can be found.
	 * @param   boolean  $reset   True to reset the prefix with only the given lookup path.
	 *
	 * @return  void
	 *
	 * @throws  RuntimeException
	 *
	 * @since   12.1
	 */
	public static function registerPrefix($prefix, $path, $reset = false)
	{
		// Verify the library path exists.
		if (!file_exists($path))
		{
			throw new RuntimeException('Library path ' . $path . ' cannot be found.', 500);
		}

		// If the prefix is not yet registered or we have an explicit reset flag then set set the path.
		if (!isset(self::$prefixes[$prefix]) || $reset)
		{
			self::$prefixes[$prefix] = array($path);
		}
		// Otherwise we want to simply add the path to the prefix.
		else
		{
			array_unshift(self::$prefixes[$prefix], $path);
		}
	}

	/**
	 * Register a namespace to the autoloader. When loaded, namespace paths are searched in a "last in, first out" order.
	 *
	 * @param   string   $namespace  A case sensitive Namespace to register.
	 * @param   string   $path       A case sensitive absolute file path to the library root where classes of the given namespace can be found.
	 * @param   boolean  $reset      True to reset the namespace with only the given lookup path.
	 *
	 * @return  void
	 *
	 * @throws  RuntimeException
	 *
	 * @since   12.3
	 */
	public static function registerNamespace($namespace, $path, $reset = false)
	{
		// Verify the library path exists.
		if (!file_exists($path))
		{
			throw new RuntimeException('Library path ' . $path . ' cannot be found.', 500);
		}

		// If the namespace is not yet registered or we have an explicit reset flag then set the path.
		if (!isset(self::$namespaces[$namespace]) || $reset)
		{
			self::$namespaces[$namespace] = array($path);
		}

		// Otherwise we want to simply add the path to the namespace.
		else
		{
			array_unshift(self::$namespaces[$namespace], $path);
		}
	}

	/**
	 * Method to setup the autoloaders for the Joomla Framework.
	 * Since the SPL autoloaders are called in a queue we will add our explicit
	 * class-registration based loader first, then fall back on the autoloader based on conventions.
	 * This will allow people to register a class in a specific location and override platform libraries
	 * as was previously possible.
	 *
	 * @param   boolean  $enableClasses     True to enable class map based class loading (needed to auto load the Joomla core).
	 * @param   boolean  $enablePrefixes    True to enable prefix based class loading (needed to auto load the Joomla core).
	 *
	 * @return  void
	 *
	 * @since   12.3
	 */
	public static function setup($enableClasses = true, $enablePrefixes = true)
	{
		spl_autoload_register(array(__CLASS__, 'loadByPsr0'));

		self::registerNamespace('Joomla', __DIR__);

		if ($enableClasses)
		{
			// Register the class map based autoloader.
			spl_autoload_register(array(__CLASS__, 'load'));
		}

		if ($enablePrefixes)
		{
			// Register the prefix autoloader.
			spl_autoload_register(array(__CLASS__, '_autoload'));
		}
	}

	/**
	 * Method to autoload classes that are namespaced to the PSR-0 standard.
	 *
	 * @param   string  $class  The fully qualified class name to autoload.
	 *
	 * @return  boolean  True on success, false otherwise.
	 *
	 * @since   13.1
	 */
	public static function loadByPsr0($class)
	{
		// Remove the root backslash if present.
		if ($class[0] == '\\')
		{
			$class = substr($class, 1);
		}

		// Find the location of the last NS separator.
		$pos = strrpos($class, '\\');

		// If one is found, we're dealing with a NS'd class.
		if ($pos !== false)
		{
			$classPath = str_replace('\\', DIRECTORY_SEPARATOR, substr($class, 0, $pos)) . DIRECTORY_SEPARATOR;
			$className = substr($class, $pos + 1);
		}
		// If not, no need to parse path.
		else
		{
			$classPath = null;
			$className = $class;
		}

		$classPath .= str_replace('_', DIRECTORY_SEPARATOR, $className) . '.php';

		// Loop through registered namespaces until we find a match.
		foreach (self::$namespaces as $ns => $paths)
		{
			if (strpos($class, $ns) === 0)
			{
				// Loop through paths registered to this namespace until we find a match.
				foreach ($paths as $path)
				{
					$classFilePath = $path . DIRECTORY_SEPARATOR . $classPath; 

					if (file_exists($classFilePath) && !class_exists($class, false))
					{
						return (bool) include_once $classFilePath;
					}
				}
			}
		}

		return false;
	}

	/**
	 * Autoload a class based on name.
	 *
	 * @param   string  $class  The class to be loaded.
	 *
	 * @return  boolean  True if the class was loaded, false otherwise.
	 *
	 * @since   11.3
	 */
	private static function _autoload($class)
	{
		foreach (self::$prefixes as $prefix => $lookup)
		{
			$chr = strlen($prefix) < strlen($class) ? $class[strlen($prefix)] : 0;

			if (strpos($class, $prefix) === 0 && ($chr === strtoupper($chr)))
			{
				return self::_load(substr($class, strlen($prefix)), $lookup);
			}
		}

		return false;
	}

	/**
	 * Load a class based on name and lookup array.
	 *
	 * @param   string  $class   The class to be loaded (without prefix).
	 * @param   array   $lookup  The array of base paths to use for finding the class file.
	 *
	 * @return  boolean  True if the class was loaded, false otherwise.
	 *
	 * @since   12.1
	 */
	private static function _load($class, $lookup)
	{
		// Split the class name into parts separated by camelCase.
		$parts = preg_split('/(?<=[a-z0-9])(?=[A-Z])/x', $class);

		// If there is only one part we want to duplicate that part for generating the path.
		$parts = (count($parts) === 1) ? array($parts[0], $parts[0]) : $parts;

		foreach ($lookup as $base)
		{
			// Generate the path based on the class name parts.
			$path = $base . '/' . implode('/', array_map('strtolower', $parts)) . '.php';

			// Load the file if it exists.
			if (file_exists($path))
			{
				return include $path;
			}
		}

		return false;
	}
}

/**
 * Intelligent file importer.
 *
 * @param   string  $path  A dot syntax path.
 *
 * @return  boolean  True on success.
 *
 * @since   11.1
 */
function jimport($path)
{
	return JLoader::import($path);
}
